hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileBlock.java

```diff
@@ -267,40 +267,40 @@ public ByteBuffer serialize() throws IOException {
     // Compress if specified.
     ByteBuffer compressedBlockData = context.getCompressor().compress(uncompressedBlockData);
     // Buffer for building block.
-    ByteBuffer buf = ByteBuffer.allocate(Math.max(
-        context.getBlockSize() * 2,
+    ByteArrayOutputStream buf = new ByteArrayOutputStream(Math.max(
+        context.getBlockSize(),
         compressedBlockData.limit() + HFILEBLOCK_HEADER_SIZE * 2));
 
     // Block header
     // 1. Magic is always 8 bytes.
-    buf.put(blockType.getMagic(), 0, 8);
+    buf.write(blockType.getMagic(), 0, 8);
     // 2. onDiskSizeWithoutHeader.
     int compressedDataSize = compressedBlockData.limit();
     int onDiskDataSizeWithHeader = HFileBlock.HFILEBLOCK_HEADER_SIZE + compressedDataSize;
     int numChecksumBytes = numChecksumBytes(onDiskDataSizeWithHeader, DEFAULT_BYTES_PER_CHECKSUM);
-    buf.putInt(compressedDataSize + numChecksumBytes);
+    HFileUtils.writeInt(buf, compressedDataSize + numChecksumBytes);
```

Comment:
If we use `DataOutputStream` to wrap the `ByteArrayOutputStream` you will get all these utilities instead of writing your own